### ISSUE TITLE
Bug report: Not all bone tracks exported in VRMa (non-humanoid bones may have keyframes too)

### ISSUE THREAD

**Issue by tdw46** [saturday06/VRM-Addon-for-Blender#584]
## Describe the bug

Non-Humanoid bones don't have keyframes exported in VRMa output.

## To Reproduce

1. Import VRM: [Test_Anim_Rig_vrm.zip](https://github.com/user-attachments/files/17195852/Test_Anim_Rig_vrm.zip)
2. Import animation: [Idle_glb-Animation.zip](https://github.com/user-attachments/files/17195860/Idle_glb-Animation.zip)
3. Apply animation to VRM rig
4. Export as VRMa, output will not have non-humanoid bone tracks. Will match: [Idle_VRM-Animation.zip](https://github.com/user-attachments/files/17195870/Idle_VRM-Animation.zip)

## Expected behavior

I expect rigs that have a lot of extra non-humanoid bones with keyframes to have those stored in the gltf animation track data.

## Screenshots

![image](https://github.com/user-attachments/assets/269c72f5-459d-4dfc-9f9d-feb43c6c5d4a) ![image](https://github.com/user-attachments/assets/9c356d9f-1691-4a49-b97c-6cc3c87dc07b)

## Desktop (please complete the following information)

- OS: Windows 11
- Blender 4.2.1 LTS
- Addon Version: Latest, 2.20.88

## Additional context

- We should embed the gltf addon preferences under "Advanced" in the VRM export preferences to mitigate these kinds of issues for power users.
- In this case, we need to include the non-humanoid bones in the VRMA exported JSON. So we'll need an advanced option for the VRMA exporter

**Comment by tdw46** [comment]
This commit ([bd5b2d3](https://github.com/saturday06/VRM-Addon-for-Blender/commit/bd5b2d3d0aaa9819969f4721660439a533ec5849)) doesn't address my needs. See second bullet under "Additional Context" where I mention these changes need to be implemented into VRMA exports-- That said, I'm working on [a pull request](https://github.com/tdw46/VRM-Addon-for-Blender-Beyond/tree/tdw46-vrma-non-humanoid-support) to address this.

**Comment by tdw46** [comment]
FYI, this is fixed [here](https://github.com/saturday06/VRM-Addon-for-Blender/pull/733)

**Comment by blancgoat** [comment]
Hello, I am someone who is very interested in this issue.    

First of all, thank you for your hard work. I have a question I'd like to ask.  
Does this PR allow inserting animations into bones other than those defined in VRM 1.0?  

I would appreciate it if you could answer my question.  

Below, I have described my situation and test results. Please take a look.  

---

### Test Procedure  
The test was conducted on commit 1f4ddc6bd8b1e204d1a6968d1264fbae2ef5c7d6  
and blender version 4.2.7

1. Import an FBX file.  
2. Import an .anim compatible with the armature and store it in an action.  
3. Use this add-on to map the armature to VRM 1.0, check **Allow Non-Humanoid Rig**, and export it as a .VRM.  
4. Export a .VRMA file as well.  
5. Open a new Blender project.  
    Import the VRM file and then import the VRMA file into the armature.  


### My Expectation  

I expected animations to be present on non-humanoid bones, but they were not.  
I am unsure whether this is because the VRMA file itself lacks the data or because of an issue with the VRMA importer.  
(Blender is the only tool I have for testing VRMA files.)  


### My Bone Structure  
```
hips
├─ spine - chest ....
└─ waist - leg_l,leg_r ....
```
the standard humanoid rig does not have a **waist** bone, so there is no way to map it.  
Since my .anim file contains animation data for both hip and waist, I was expecting this PR to resolve the issue.

---

Since the situation you resolved may be different from mine, I wanted to ask this question. 

Thank you so much for reading until the end🙌

**Comment by tdw46** [comment]
> Hello, I am someone who is very interested in this issue.
> 
> First of all, thank you for your hard work. I have a question I'd like to ask. Does this PR allow inserting animations into bones other than those defined in VRM 1.0?

@blancgoat Yes, this modification to the exporter will resolve your issue and allow exporting animation tracks that are not part of the Unity Humanoid standard. However, because such "non-humanoid" bones are not in the VRM / [VRMA spec](https://github.com/vrm-c/vrm-specification/tree/master/specification/VRMC_vrm_animation-1.0), most VRM apps with VRMA importers will not import animations for these non-humanoid bones. 

My company has built a custom importer for our website that supports this. We plan to submit a feature request to the official spec, so when we do it would be helpful if you (and any friends you convince) vote for the feature request! I'll share a link to the request when I've submitted it. (After this update is merged to main, I am making a few tweaks to my code here for some workarounds to support scale and transform tracks on all bones without breaking spec). This is the fastest way to ensure VRM apps support this feature.

**Comment by blancgoat** [comment]
Thank you for your response!  
Please ignore the test process above. The issue occurred because I applied the add-on incorrectly.

After correctly applying the commit, I confirmed that more options are available when exporting VRMA.

The good news is that I confirmed the animation is applied even to non-human bones!
Unfortunately in my case, the animation appears somewhat distorted upon application. 
Also, I expected it to work as usual when unchecking all the option checkboxes in VRMA, but that wasn’t the case. (It seems that non-human bones are twisted at certain angles.)  

(Please note that I am conducting VRMA tests only using Blender)

This could be due to a mistake in my testing method.
I'll wait for the official add-on update. Thank you!

**Comment by tdw46** [comment]
Ahh @blancgoat the latest commit is broken. Please check back later, I’ll fix things and ping you when it should be working (this week sometime)

**Comment by saturday06** [comment]
@tdw46  Thank you for your patience. I am back in development.

~~Exporting rotation of non-humanoid skeletons should be enabled by default, with no settings.~~

~~This is to keep the configuration as simple as possible.~~  (Edit in 2025-03-28: https://github.com/saturday06/VRM-Addon-for-Blender/issues/584#issuecomment-2760649745 )

However, if you are going to break the spec and add translation and scale tracks to vrma, please make sure that the export extension is glb. Because it is not vrma and it will confuse applications that read vrma.

Please submit the specs you want here
https://github.com/vrm-c/vrm-specification/issues

If approved, change the extension from glb to vrma.

**Comment by saturday06** [comment]
If you want to standardize a non-standard specification, it is more reliable to submit a Pull Request to UniVRM instead of Blender's add-on.

UniVRM is the official reference implementation, but it includes experimental features like cloth simulation.

https://github.com/vrm-c/UniVRM

I am working hard to make sure that Blender add-ons are consistent with UniVRM's behavior. It is not quite complete yet. So there is a lot of work to be done prior to putting in a non-standard spec.

**Comment by saturday06** [comment]
I too would like to have translaion and scale animations. I would also like to have ik animation.

**Comment by tdw46** [comment]
@saturday06 thank you! Actually my latest string of commits is to implement a workaround to include scale and translation in nodes that aren’t pointed to by the VRMA extension! This doesn’t break spec. Additionally, I have an option to generate non-humanoid bones that match hierarchy of humanoid skeleton perfectly to store all the translation and scale tracks on any humanoid bones. This is non-destructive and only happens at export and is then undone. To be sure other apps are happy, I tested a basic asset that does this in UniVRM and three-VRM and both performed well. More details soon. I am working on exporter UI options to make this intuitive, with defaults to rotation only.

**Comment by saturday06** [comment]
Thank you.

Please put this feature behind “Advanced Options” so I can ask for it. I think it would effectively be a Blender or your company's importer exclusive feature, similar to the option of unlimited bone weights available only in Unreal Engine.

Once it is finalized as a VRM Consortium specification, I can enable it by default.

**Comment by tdw46** [comment]
Perfect, that matches my plans! This also adds some booleans to the VRMA exporter panel. Would you like those hidden under an advanced options group there, as well?

**Comment by saturday06** [comment]
I would like to do basically the same thing, depending on the functionality.

I want to reduce the number of settings that a novice user sees, so I hide them in “Advanced Options” and enable them by default when the VRM Consortium specs become available.

~~However, for example, “Export non-humanoid rotation animation” is supported by UniVRM, so I think it is better to enable it by default and not create a setting for it.~~  (Edit in 2025-03-28: https://github.com/saturday06/VRM-Addon-for-Blender/issues/584#issuecomment-2760649745 )

**Comment by saturday06** [comment]
It is important that the result be the same as UniVRM. Settings that do not do so are hidden in “Advanced Options.”

**Comment by saturday06** [comment]
I misunderstood. UniVRM actually didn't support every type of animation for Non-Humanoid bones within VRM Animation files. Because of this, any extensions concerning Non-Humanoid bones had to be hidden under "Advanced Options".
